### PR TITLE
fix(files): Provide default file action for file entry name (on click action)

### LINKED ISSUE #44855: [Bug]: Hash in filename (e.g. #foo.docx) breaks web interface download
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Downloading a file that has a hash in the name doesn't work anymore when clicking on the file row.
Context menu / 3 dot menu -> download still works

### Steps to reproduce

1. Upload file `#foo.docx`
2. Click on file row to open it (in this case download it as file)
3. A file with the correct name is downloaded. It's corrupt because it has the following content:

> This is the WebDAV interface. It can only be accessed by WebDAV clients such as the Nextcloud desktop sync client

This is because the UI tries to download the folder, not the file.
This might be because it ignores everything in the URL that comes after `#` 

Note: This behavior can't be reproduced for PDFs when the PDF viewer is active.

### Expected behavior

1. Upload file '#foo.docx'
2. Click on file row to open (in this case download) it
3. A file with the correct name is downloaded. It's a copy of the uploaded file.

### Installation method

Community Docker image

### Nextcloud Server version

28

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### LINKED ISSUE #45678: [Bug]: Wrong tooltip (hover) on file and folder names in Nextcloud Files
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

In Nextcloud 28 and 29 (not in Nextcloud 27), when you arrive in a folder, you'll notice that the tooltips are wrong at first display when hovering files and folders in an existing folder. For both files and folders, the tooltip is "Download file {name of file or folder}". Then, if you refresh the list, tooltips are finally fine.

[Capture vidéo du 05-06-2024 16:22:12.webm](https://github.com/nextcloud/server/assets/33763786/0c887d7d-d0c1-4b40-8539-bb757cc334cf)

### Steps to reproduce

1. Create a folder at root in a NC 28 or 29 instance 
2. Add files and create folders inside it
3. Get inside it by clicking on it
4. Notice the wrong tooltips by hovering files and subfolders 
5. Refresh the list (by clicking on the breadcrumb or by using the browser's refresh button) and see good ones ("View" for files and "Open folder" for folders)

### Expected behavior

We should **always** have the following tooltips : 
- "View" for files
- "Open folder" for folders

### Installation method

Community Manual installation with Archive

### Nextcloud Server version

28

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.2

### Web server

Nginx

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Updated from a MINOR version (ex. 22.1 to 22.2)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [X] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "nc28.test.mydomain.com"
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "dbtype": "mysql",
    "version": "28.0.6.1",
    "overwrite.cli.url": "https:\/\/nc28.test.mydomain.com",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "mysql.utf8mb4": true,
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "log_type": "file",
    "logfile": "\/var\/log\/nextcloud\/nc28.test.mydomain.com\/nextcloud.log",
    "redis": {
        "host": "***REMOVED SENSITIVE VALUE***",
        "password": "***REMOVED SENSITIVE VALUE***",
        "port": 6379,
        "dbindex": 1,
        "timeout": 0
    },
    "skeletondirectory": "",
    "logtimezone": "Europe\/Paris",
    "updater.release.channel": "beta",
    "mail_smtpmode": "smtp",
    "mail_smtpsecure": "ssl",
    "mail_from_address": "***REMOVED SENSITIVE VALUE***",
    "mail_domain": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
    "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
    "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpauthtype": "PLAIN",
    "default_language": "fr",
    "default_phone_region": "FR",
    "defaultapp": "files",
    "trashbin_retention_obligation": "20,40",
    "versions_retention_obligation": "auto,40",
    "appstoreurl": "https:\/\/appstore.test.mydomain.com\/api\/v1",
    "check_for_working_wellknown_setup": false,
    "quota_include_external_storage": false,
    "cron_log": true,
    "has_internet_connection": true,
    "updatechecker": true,
    "appstoreenabled": true,
    "filelocking.enabled": true,
    "session_keepalive": true,
    "knowledgebaseenabled": true,
    "allow_user_to_change_display_name": true,
    "enable_previews": true,
    "enable_avatars": true,
    "auth.bruteforce.protection.enabled": true,
    "loglevel": 1,
    "log_rotate_size": 104857600,
    "mail_smtpauth": 1,
    "mail_smtpport": 465,
    "session_lifetime": 86400,
    "remember_login_cookie_lifetime": 1296000,
    "preview_max_filesize_image": 50,
    "activity_expire_days": 120,
    "memcache.local": "\\OC\\Memcache\\APCu",
    "memcache.locking": "\\OC\\Memcache\\Redis",
    "memcache.distributed": "\\OC\\Memcache\\Redis",
    "app_install_overwrite": [
        "admin_audit",
        "user_ldap",
        "workspace"
    ],
    "ldapProviderFactory": "OCA\\User_LDAP\\LDAPProviderFactory",
    "maintenance": false,
    "maintenance_window_start": 1
}
```

### List of activated Apps

```shell
Enabled:
 - activity: 2.20.0
 - admin_audit: 1.18.0
 - assistant: 1.0.8
 - bbb: 2.6.0
 - calendar: 4.7.2
 - circles: 28.0.0
 - cloud_federation_api: 1.11.0
 - comments: 1.18.0
 - contacts: 5.5.3
 - contactsinteraction: 1.9.0
 - context_chat: 2.1.1
 - dashboard: 7.8.0
 - dav: 1.29.1
 - deck: 1.12.2
 - electronicsignatures: 3.0.0
 - federatedfilesharing: 1.18.0
 - federation: 1.18.0
 - files: 2.0.0
 - files_pdfviewer: 2.9.0
 - files_reminders: 1.1.0
 - files_sharing: 1.20.0
 - files_trashbin: 1.18.0
 - files_versions: 1.21.0
 - firstrunwizard: 2.17.0
 - groupfolders: 16.0.6
 - impersonate: 1.15.0
 - logreader: 2.13.0
 - lookup_server_connector: 1.16.0
 - mail: 3.6.0
 - nextcloud_announcements: 1.17.0
 - notifications: 2.16.0
 - oauth2: 1.16.3
 - onlyoffice: 9.2.0
 - password_policy: 1.18.0
 - photos: 2.4.0
 - privacy: 1.12.0
 - provisioning_api: 1.18.0
 - recommendations: 2.0.0
 - related_resources: 1.3.0
 - richdocuments: 8.3.5
 - serverinfo: 1.18.0
 - settings: 1.10.1
 - sharebymail: 1.18.0
 - support: 1.11.1
 - systemtags: 1.18.0
 - text: 3.9.1
 - theming: 2.3.0
 - twofactor_backupcodes: 1.17.0
 - updatenotification: 1.18.0
 - user_ldap: 1.19.0
 - user_status: 1.8.1
 - viewer: 2.2.0
 - weather_status: 1.8.0
 - workflowengine: 2.10.0
 - workspace: 3.2.0-beta1
Disabled:
 - bruteforcesettings
 - encryption
 - files_external
 - survey_client: 1.16.0
 - suspicious_login
 - twofactor_totp
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```shell
Can be provided if necessary
```

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by susnux**
* Resolves: https://github.com/nextcloud/server/issues/44855
* Resolves: https://github.com/nextcloud/server/issues/45678

## Summary

This fixes two issues with the files actions:
1. Current the actions were handled by the FileListEntryActions component, and then weirdly accessed through the shadow DOM like `$parent.$refs.something` this is not reactive and leads to situations where the default actions was not correctly injected -> Tooltip was wrong
2. We had two places where the download action was defined, first the download action and second as a fallback in the name component. So this is consolidated into only one place: The download file action. Meaning the download file action is now a default action. And if this is not injected in the name component, then the user can not download -> no fallback needed.
3. (basically 2.1) the fallback download action was not using the encoded source, thus e.g. names with # were not correctly downloaded.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/files/src/components/FileEntry/FileEntryName.vue [hunk 1]
```diff
@@ -280,20 +274,20 @@ export default defineComponent({
...
 				this.$nextTick(() => {
-					this.$refs.basename?.focus()
+					const nameContainter = this.$refs.basename as HTMLElement | undefined
+					nameContainter?.focus()
 				})
 			} catch (error) {
 				logger.error('Error while renaming file', { error })
-				this.source.rename(oldName)
-				this.$refs.renameInput?.focus()
+				this.startRenaming()
```
Comment (skjnldsv) [inline apps/files/src/components/FileEntry/FileEntryName.vue:282]: Seems a bit unrelated? :see_no_evil:

**Review by skjnldsv — APPROVED** [review]

**Comment by skjnldsv** [comment]
Yeah, the old legacy link/action was too weird in the end.
It's much cleaner to only rely on the provided file actions, you're 100% right!

File: apps/files/src/components/FileEntry/FileEntryName.vue (hunk 1 above)
Comment (susnux) [reply apps/files/src/components/FileEntry/FileEntryName.vue:282]: Right, adjusted the code here

**Review by sorbaugh — APPROVED** [review]

**Pull request merged**